docs/secrets.md

```diff
@@ -0,0 +1,66 @@
+Notification endpoints will require credentials in order to communicate with the various external services.
+
+We do not want users to include those credentials in plain text into Flux scripts. We need to provide a mechanism for users load secrets from a safe storage mechanism and provide those secrets to the endpoints.
+
+**Backing Store:** : There are many backing store possibilities, which need not be discussed here.  Any store that provides secure access for key/value lookup may be used by providing an implementation of the KeyLookup  interface in the flux repository.  
+
+**Fine Grained Authorization**:  keys will generally not be accessible in the flux language.  A secret-enabled function will accept meta-data about the secret (a key) and use it to look up the stored value in memory. The lookup will error if the key is not found or if the user is not authorized to use that key.  
+
+# **Design**
+
+## Internal Functions
+A function that requires secret lookup must: 
+
+- Be built-in to the language
+- Register an ExternalDependencies struct  (note: it will be straightforward and advantageous to refactor the `InjectFromDependencies` and similar functions into this pattern as well).   
+- Not return secrets under any circumstances (e.g. in error messages) 
+
+The external interface for secret lookup is: 
+```Go
+func LookupRequest(keyObject values.Object) SecretRequest
```

Comment:
Can you remove the `values.Object` type from this interface? Looking up secrets should not be coupled to how we represent values inside of the Flux interpreter. Can you change this function to accept the raw arguments it needs instead of extracting them from a `values.Object`?